cmd/awsDataGather/main.go

```diff
@@ -0,0 +1,123 @@
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/red-hat-storage/ocs-osd-deployer/pkg/aws"
+	"github.com/red-hat-storage/ocs-osd-deployer/utils"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+func main() {
+	ctrl.SetLogger(zap.New(zap.UseDevMode(true)))
+	log := ctrl.Log.WithName("main")
+
+	namespace, found := os.LookupEnv("NAMESPACE")
+	if !found {
+		fmt.Fprintf(os.Stderr,
+			"NAMESPACE environment variable not found\n")
+		os.Exit(1)
+	}
+	podName, found := os.LookupEnv("POD_NAME")
+	if !found {
+		fmt.Fprintf(os.Stderr,
+			"POD_NAME environment variable not found\n")
+		os.Exit(1)
+	}
+
+	// We need the deployment name to use as an owner reference.
+	// However, the deployment name cannot be passed as an environment variable,
+	// so we derive it from the pod name.
+	deploymentName, err := utils.DeploymentNameFromPodName(podName)
+	if err != nil {
+		fmt.Fprintf(os.Stderr,
+			"Could not determine deployment name from pod name (%s)",
+			podName)
+		os.Exit(1)
+	}
+
+	log.Info("Setting up k8s client")
+	var options client.Options
+	options.Scheme = runtime.NewScheme()
+	utilruntime.Must(clientgoscheme.AddToScheme(options.Scheme))
+
+	k8sClient, err := client.New(config.GetConfigOrDie(), options)
+	if err != nil {
+		log.Error(err, "error creating client")
+		os.Exit(1)
+	}
+
+	// This will later be used as the OwnerReference for the data ConfigMap
+	var deployment appsv1.Deployment
+	err = utils.Retry(5, 1*time.Second, func() error {
+		err := k8sClient.Get(context.Background(), client.ObjectKey{
+			Name:      deploymentName,
+			Namespace: namespace,
+		}, &deployment)
+		return err
+	})
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Failed to find deployment '%s' in namespace '%s'", deploymentName, namespace))
+		os.Exit(1)
+	}
+
+	log.Info("Gathering AWS data")
+	if err := gatherAndSaveData(aws.IMDSv1Server, deployment, k8sClient); err != nil {
+		log.Error(err, "Failed to gather AWS data")
+		os.Exit(1)
+	}
+
+	log.Info("AWS data gathering successfully completed!")
+
+	// Set up signal handler, so that a clean up procedure will be run if the pod running this code is deleted.
+	// This code is run as part of a deployment that is pushed to Kubernetes through an OLM Bundle.
+	// At this time, OLM Bundles do not support restartPolicy: OnFailure,
+	// so terminating this program will cause the pod to keep getting restarted and needlessly running.
+	// To avoid this, a signal handler is set up to only terminate the program when the pod is getting deleted.
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, syscall.SIGTERM)
+	<-signalChan
```

Comment:
It is used to block completion of main until the pod receives the SIGTERM signal during deletion. You may have missed the comment explaining the reason in the lines above this:
```
	// Set up signal handler, so that a clean up procedure will be run if the pod running this code is deleted.
	// This code is run as part of a deployment that is pushed to Kubernetes through an OLM Bundle.
	// At this time, OLM Bundles do not support restartPolicy: OnFailure,
	// so terminating this program will cause the pod to keep getting restarted and needlessly running.
	// To avoid this, a signal handler is set up to only terminate the program when the pod is getting deleted.
```

Reply:
@renan-campos I am not sure I understand the issue you are trying to solve here. If the pod is getting deleted the program will terminate anyway. Why do we need to track this via an OS signal?

Reply:
We only want the program to terminate when the pod is deleted. Having this signal handler here ensures that happens. If main wasn't blocked waiting for the signal, then it would complete and the pod would endlessly restart

Reply:
After discussion in-call discussion: this signal handling won't be needed after switching to a looping structure for main.